# In this notebook we want to preprocess data better so after preprocessing we don't end up with duplicate rows with the same county of the same date.

1. produce reduced dataset with only columns we need

In [1]:
import pandas as pd

In [74]:
data = pd.read_csv("social_dist_all_trimmed_new.csv", nrows = 0)
data.drop(['Unnamed: 0'], axis = 1, inplace = True)
data = data[['date_range_start', 'state', 'state_code', 'cnamelong',\
             'county_code', 'device_count', 'completely_home_device_count']]
data.to_csv("social_dist_all_reduced.csv", index = False)

In [3]:
for chunk in pd.read_csv("social_dist_all_trimmed_new.csv", chunksize=1000):
    chunk.drop(['Unnamed: 0'], axis = 1, inplace = True)
    chunk = chunk[['date_range_start', 'state', 'state_code', 'cnamelong',\
                   'county_code', 'device_count', 'completely_home_device_count']]
    chunk.to_csv("social_dist_all_reduced.csv", index = False, header = False, mode = 'a')

In [15]:
# this is fine.
data = pd.read_csv("social_dist_all_trimmed_new.csv", nrows = 219490)

In [16]:
data.tail()

Unnamed: 0           date_range_start             date_range_end  \
219485      219485  2019-01-01T00:00:00-06:00  2019-01-02T00:00:00-06:00   
219486      219486  2019-01-01T00:00:00-06:00  2019-01-02T00:00:00-06:00   
219487      219487  2019-01-01T00:00:00-06:00  2019-01-02T00:00:00-06:00   
219488      219488  2019-01-01T00:00:00-06:00  2019-01-02T00:00:00-06:00   
219489      219489  2019-01-01T00:00:00-04:00  2019-01-02T00:00:00-04:00   

        state state_code            cnamelong  county_code  \
219485   55.0         WI       Kenosha County        55059   
219486   55.0         WI       Kenosha County        55059   
219487   55.0         WI      Langlade County        55067   
219488   55.0         WI      Marathon County        55073   
219489   72.0         PR  CanÛvanas Municipio        72029   

        origin_census_block_group  candidate_device_count  device_count  \
219485               550590013003                     175            60   
219486               550590020001                     145            56   
219487               550679605002                     125            33   
219488               550730019007                     113            39   
219489               720291001033                      60            20   

        completely_home_device_count  part_time_work_behavior_devices  \
219485                            29                                3   
219486                            17                                1   
219487                            16                                1   
219488                            11                                2   
219489                             6                                1   

        full_time_work_behavior_devices  delivery_behavior_devices  \
219485                                4                          1   
219486                                1                          1   
219487                                1                          1   
219488                                1                          1   
219489                                1                          1   

        median_home_dwell_time  median_non_home_dwell_time  \
219485                     797                           0   
219486                    1037                          64   
219487                     476                           0   
219488                     786                         119   
219489                     238                          56   

        median_percentage_time_home  distance_traveled_from_home  
219485                          100                         3264  
219486                           95                        11727  
219487                          100                         2978  
219488                           88                         1953  
219489                           68                         2539

In [60]:
# this gives some warning.
data = pd.read_csv("social_dist_all_trimmed_new.csv", nrows = 219491)

C:\Users\wenju\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (3,6,7,8,9,10,11,12,13,14,15,16,17) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [18]:
data.tail()

Unnamed: 0           date_range_start             date_range_end  \
219486    219486.0  2019-01-01T00:00:00-06:00  2019-01-02T00:00:00-06:00   
219487    219487.0  2019-01-01T00:00:00-06:00  2019-01-02T00:00:00-06:00   
219488    219488.0  2019-01-01T00:00:00-06:00  2019-01-02T00:00:00-06:00   
219489    219489.0  2019-01-01T00:00:00-04:00  2019-01-02T00:00:00-04:00   
219490         NaN           date_range_start             date_range_end   

        state  state_code            cnamelong  county_code  \
219486     55          WI       Kenosha County        55059   
219487     55          WI      Langlade County        55067   
219488     55          WI      Marathon County        55073   
219489     72          PR  CanÛvanas Municipio        72029   
219490  state  state_code            cnamelong  county_code   

        origin_census_block_group  candidate_device_count  device_count  \
219486               550590020001                     145            56   
219487               550679605002                     125            33   
219488               550730019007                     113            39   
219489               720291001033                      60            20   
219490  origin_census_block_group  candidate_device_count  device_count   

        completely_home_device_count  part_time_work_behavior_devices  \
219486                            17                                1   
219487                            16                                1   
219488                            11                                2   
219489                             6                                1   
219490  completely_home_device_count  part_time_work_behavior_devices   

        full_time_work_behavior_devices  delivery_behavior_devices  \
219486                                1                          1   
219487                                1                          1   
219488                                1                          1   
219489                                1                          1   
219490  full_time_work_behavior_devices  delivery_behavior_devices   

        median_home_dwell_time  median_non_home_dwell_time  \
219486                    1037                          64   
219487                     476                           0   
219488                     786                         119   
219489                     238                          56   
219490  median_home_dwell_time  median_non_home_dwell_time   

        median_percentage_time_home  distance_traveled_from_home  
219486                           95                        11727  
219487                          100                         2978  
219488                           88                         1953  
219489                           68                         2539  
219490  median_percentage_time_home  distance_traveled_from_home

Now we know why some of the processing below need to deal with invalid rows: in some intervals in the original data we have the header instead of the actual data. (must be when we insert the data into this aggregated dataframe we included the header, which is not what we want.)

This shows the weakness of dynamically typed language. In python we have no way to enforce the schema of the dataframe: we can't enforce "device_count" column only takes in integer datatype. Instead what Pandas does is to infer the datatype by reading all the rows, which takes a long time.. 

Now I have a theory why we have duplicated rows: maybe we have inconsistent data types: "55" (string) != 55 (integer), so our groupby didn't work as expected..

In [63]:
data.isnull().sum()

Unnamed: 0                          1
date_range_start                    0
date_range_end                      0
state                              13
state_code                         13
cnamelong                          13
county_code                         0
origin_census_block_group           0
candidate_device_count              0
device_count                        0
completely_home_device_count        0
part_time_work_behavior_devices     0
full_time_work_behavior_devices     0
delivery_behavior_devices           0
median_home_dwell_time              0
median_non_home_dwell_time          0
median_percentage_time_home         0
distance_traveled_from_home         0
dtype: int64

In [64]:
data[data['cnamelong'].isnull()]

Unnamed: 0           date_range_start             date_range_end  \
93214      93214.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
96657      96657.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
125011    125011.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
132920    132920.0  2019-01-01T00:00:00-09:00  2019-01-02T00:00:00-09:00   
133936    133936.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
158801    158801.0  2019-01-01T00:00:00-09:00  2019-01-02T00:00:00-09:00   
169207    169207.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
174123    174123.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
180175    180175.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
190133    190133.0  2019-01-01T00:00:00-09:00  2019-01-02T00:00:00-09:00   
207541    207541.0  2019-01-01T00:00:00-09:00  2019-01-02T00:00:00-09:00   
215950    215950.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
219452    219452.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   

       state state_code cnamelong county_code origin_census_block_group  \
93214    NaN        NaN       NaN       46102              461029405003   
96657    NaN        NaN       NaN       46102              461029409003   
125011   NaN        NaN       NaN       46102              461029409002   
132920   NaN        NaN       NaN        2158               21580001003   
133936   NaN        NaN       NaN       46102              461029409001   
158801   NaN        NaN       NaN        2158               21580001004   
169207   NaN        NaN       NaN       46102              461029405001   
174123   NaN        NaN       NaN       46102              461029408002   
180175   NaN        NaN       NaN       46102              461029408001   
190133   NaN        NaN       NaN        2158               21580001002   
207541   NaN        NaN       NaN       02158              021580001001   
215950   NaN        NaN       NaN       46102              461029408003   
219452   NaN        NaN       NaN       46102              461029405002   

       candidate_device_count device_count completely_home_device_count  \
93214                     185           59                           31   
96657                     171           43                           19   
125011                    129           47                           24   
132920                    107           32                           10   
133936                    182           43                           13   
158801                    213           38                           15   
169207                    184           52                           26   
174123                    136           25                           16   
180175                     98           28                           10   
190133                     48           12                            1   
207541                    174           42                           18   
215950                    257           64                           35   
219452                    186           67                           31   

       part_time_work_behavior_devices full_time_work_behavior_devices  \
93214                                1                               1   
96657                                1                               1   
125011                               2                               1   
132920                               1                               1   
133936                               1                               1   
158801                               1                               1   
169207                               1                               1   
174123                               4                               3   
180175                               1                               1   
190133                               1                               1   
207541                     

In [68]:
data[data['county_code'] == 46102]

Unnamed: 0           date_range_start             date_range_end  \
93214      93214.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
96657      96657.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
125011    125011.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
133936    133936.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
169207    169207.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
174123    174123.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
180175    180175.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   

       state state_code cnamelong county_code origin_census_block_group  \
93214    NaN        NaN       NaN       46102              461029405003   
96657    NaN        NaN       NaN       46102              461029409003   
125011   NaN        NaN       NaN       46102              461029409002   
133936   NaN        NaN       NaN       46102              461029409001   
169207   NaN        NaN       NaN       46102              461029405001   
174123   NaN        NaN       NaN       46102              461029408002   
180175   NaN        NaN       NaN       46102              461029408001   

       candidate_device_count device_count completely_home_device_count  \
93214                     185           59                           31   
96657                     171           43                           19   
125011                    129           47                           24   
133936                    182           43                           13   
169207                    184           52                           26   
174123                    136           25                           16   
180175                     98           28                           10   

       part_time_work_behavior_devices full_time_work_behavior_devices  \
93214                                1                               1   
96657                                1                               1   
125011                               2                               1   
133936                               1                               1   
169207                               1                               1   
174123                               4                               3   
180175                               1                               1   

       delivery_behavior_devices median_home_dwell_time  \
93214                          1                    469   
96657                          1                    410   
125011                         1                    799   
133936                         1                    145   
169207                         1                    380   
174123                         3                    369   
180175                         1                    403   

       median_non_home_dwell_time median_percentage_time_home  \
93214                           0                         100   
96657                           0                         100   
125011                          0                         100   
133936                         68                          55   
169207                          0                         100   
174123                          0                         100   
180175                         13                          99   

       distance_traveled_from_home  
93214                         1094  
96657                          228  
125011                        4537  
133936                        2969  
169207                         318  
174123                         551  
180175                           0

In [69]:
data[data['county_code'] == '46102']

Unnamed: 0           date_range_start             date_range_end  \
215950    215950.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   
219452    219452.0  2019-01-01T00:00:00-07:00  2019-01-02T00:00:00-07:00   

       state state_code cnamelong county_code origin_census_block_group  \
215950   NaN        NaN       NaN       46102              461029408003   
219452   NaN        NaN       NaN       46102              461029405002   

       candidate_device_count device_count completely_home_device_count  \
215950                    257           64                           35   
219452                    186           67                           31   

       part_time_work_behavior_devices full_time_work_behavior_devices  \
215950                               4                               4   
219452                               1                               1   

       delivery_behavior_devices median_home_dwell_time  \
215950                         2                    673   
219452                         1                    227   

       median_non_home_dwell_time median_percentage_time_home  \
215950                          0                         100   
219452                         21                          96   

       distance_traveled_from_home  
215950                         473  
219452                         372

There are rows in the original dataset that have null county names. Don't know why this happens? 

1 way is to include these 12 rows in our analysis and another way is to drop them.

I am choosing to drop these rows. This excludes county with FIPS code of 46102 and 2158 in our analysis, which I don't think is that much of a deal. Just 2 counties.

In [81]:
data = pd.read_csv("social_dist_all_reduced.csv", nrows = 219491)

C:\Users\wenju\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (1,4,5,6) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [82]:
data.tail()

date_range_start  state  state_code            cnamelong  \
219486  2019-01-01T00:00:00-06:00     55          WI       Kenosha County   
219487  2019-01-01T00:00:00-06:00     55          WI      Langlade County   
219488  2019-01-01T00:00:00-06:00     55          WI      Marathon County   
219489  2019-01-01T00:00:00-04:00     72          PR  CanÛvanas Municipio   
219490           date_range_start  state  state_code            cnamelong   

        county_code  device_count  completely_home_device_count  
219486        55059            56                            17  
219487        55067            33                            16  
219488        55073            39                            11  
219489        72029            20                             6  
219490  county_code  device_count  completely_home_device_count

In [83]:
data.dtypes

date_range_start                object
state                           object
state_code                      object
cnamelong                       object
county_code                     object
device_count                    object
completely_home_device_count    object
dtype: object

In [84]:
data['device_count'] = pd.to_numeric(data['device_count'], errors='coerce')

In [85]:
data['completely_home_device_count'] = pd.to_numeric(data['completely_home_device_count'], errors='coerce')

In [86]:
data['county_code'] = pd.to_numeric(data['county_code'], errors='coerce')

In [87]:
data['state'] = pd.to_numeric(data['state'], errors='coerce')

In [88]:
data.tail()

date_range_start  state  state_code            cnamelong  \
219486  2019-01-01T00:00:00-06:00   55.0          WI       Kenosha County   
219487  2019-01-01T00:00:00-06:00   55.0          WI      Langlade County   
219488  2019-01-01T00:00:00-06:00   55.0          WI      Marathon County   
219489  2019-01-01T00:00:00-04:00   72.0          PR  CanÛvanas Municipio   
219490           date_range_start    NaN  state_code            cnamelong   

        county_code  device_count  completely_home_device_count  
219486      55059.0          56.0                          17.0  
219487      55067.0          33.0                          16.0  
219488      55073.0          39.0                          11.0  
219489      72029.0          20.0                           6.0  
219490          NaN           NaN                           NaN

In [89]:
data.dropna(inplace = True)

In [90]:
data.tail()

date_range_start  state state_code            cnamelong  \
219485  2019-01-01T00:00:00-06:00   55.0         WI       Kenosha County   
219486  2019-01-01T00:00:00-06:00   55.0         WI       Kenosha County   
219487  2019-01-01T00:00:00-06:00   55.0         WI      Langlade County   
219488  2019-01-01T00:00:00-06:00   55.0         WI      Marathon County   
219489  2019-01-01T00:00:00-04:00   72.0         PR  CanÛvanas Municipio   

        county_code  device_count  completely_home_device_count  
219485      55059.0          60.0                          29.0  
219486      55059.0          56.0                          17.0  
219487      55067.0          33.0                          16.0  
219488      55073.0          39.0                          11.0  
219489      72029.0          20.0                           6.0

In [91]:
data.dtypes

date_range_start                 object
state                           float64
state_code                       object
cnamelong                        object
county_code                     float64
device_count                    float64
completely_home_device_count    float64
dtype: object

Now our county_code and device_count are all of the same datatype (float), so when we do groupby and aggregate, it should be fine.

In [119]:
from datetime import datetime

In [204]:
# enforce a datatype schema so groupby functions properly.
aggregatedData = pd.read_csv("social_dist_all_reduced.csv", nrows = 0, \
                             dtype = {'date_range_start': 'datetime64[ns]',\
                                        'state': 'float64', \
                                        'state_code': 'str', \
                                        'cnamelong': 'str', \
                                        'county_code': 'float64', \
                                        'device_count': 'float64', \
                                        'completely_home_device_count': 'float64'})

In [205]:
aggregatedData

Empty DataFrame
Columns: [date_range_start, state, state_code, cnamelong, county_code, device_count, completely_home_device_count]
Index: []

In [206]:
aggregatedData.dtypes

date_range_start                datetime64[ns]
state                                  float64
state_code                              object
cnamelong                               object
county_code                            float64
device_count                           float64
completely_home_device_count           float64
dtype: object

In [207]:
counter = 0
for chunk in pd.read_csv("social_dist_all_reduced.csv", chunksize = 1000000):
    # convert chunk's columns to the right datatype.
    chunk['device_count'] = pd.to_numeric(chunk['device_count'], errors='coerce')
    chunk['completely_home_device_count'] = pd.to_numeric(chunk['completely_home_device_count'], errors='coerce')
    chunk['county_code'] = pd.to_numeric(chunk['county_code'], errors='coerce')
    chunk['state'] = pd.to_numeric(chunk['state'], errors='coerce')
    # drop the null rows (get rid of the row that is header)
    # also gets rid of 2 counties whose FIPS code is null.
    chunk.dropna(inplace = True)
    # convert string column to datetime object
    chunk['date_range_start'] = chunk['date_range_start'].apply(lambda x: datetime.strptime(x[0:10], '%Y-%m-%d'))
    
    # groupby and aggregate data of the same county of the same date.
    aggregatedData = aggregatedData.append(chunk, ignore_index = True)
    aggregatedData = aggregatedData.groupby(['date_range_start', 'state', 'state_code',\
                                             'cnamelong', 'county_code']).agg(device_count = ('device_count', 'sum'),\
                                                                              completely_home_device_count = ('completely_home_device_count', 'sum')).reset_index()
    print(counter) # track progress
    print(aggregatedData.shape[0]) # how large is our data right now
    counter += 1

C:\Users\wenju\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (1,4,5,6) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


0
16078
1
31642
2
45136
3
61012
4
74182
5
90190
6
103226
7
119274
8
132264
9
148347
10
163825
11
177399
12
193246
13
206447
14
222430
15
235492
16
251533
17
264531
18
280606
19
295997
20
309664
21
325489
22
338709
23
354688
24
367752
25
383789
26
396787
27
412857
28
428225
29
441896
30
457720
31
470943
32
486931
33
499984
34
516029
35
529020
36
545091
37
560551
38
574151
39
590011
40
603191
41
619198
42
632236
43
648282
44
661988
45
677344
46
692973
47
706397
48
722305
49
735441
50
751465
51
764477
52
780531
53
795342
54
809585
55
825311
56
838642
57
854583
58
867690
59
883726
60
896726
61
912784
62
928123
63
941832
64
957699
65
970862
66
986867
67
999887
68
1015942
69
1031267
70
1044986
71
1060864
72
1074027
73
1090036
74
1103047
75
1119100
76
1134233
77
1148166
78
1163993
79
1177214
80
1193181
81
1206260
82
1222307
83
1235299
84
1251375
85
1266898
86
1280432
87
1296308
88
1309465
89
1325449
90
1338475
91
1354518
92
1369029
93
1383565
94
1399231
95
1412621
96
1428537
97
1441670
98
145

Last time we did this we got 4 million rows, now we only get 2 million rows. Why??

In [208]:
aggregatedData

date_range_start  state state_code          cnamelong  county_code  \
0             2019-01-01    1.0         AL     Autauga County       1001.0   
1             2019-01-01    1.0         AL     Baldwin County       1003.0   
2             2019-01-01    1.0         AL     Barbour County       1005.0   
3             2019-01-01    1.0         AL        Bibb County       1007.0   
4             2019-01-01    1.0         AL      Blount County       1009.0   
...                  ...    ...        ...                ...          ...   
1889781       2020-08-08   72.0         PR  Yabucoa Municipio      72151.0   
1889782       2020-08-08   72.0         PR    Yauco Municipio      72153.0   
1889783       2020-08-08   78.0         VI   St. Croix Island      78010.0   
1889784       2020-08-08   78.0         VI    St. John Island      78020.0   
1889785       2020-08-08   78.0         VI  St. Thomas Island      78030.0   

         device_count  completely_home_device_count  
0              4708.0                        1829.0  
1             19655.0                        7717.0  
2              1570.0                         594.0  
3              1702.0                         623.0  
4              5224.0                        1901.0  
...               ...                           ...  
1889781         385.0                         122.0  
1889782         519.0                         198.0  
1889783        1083.0                         295.0  
1889784         125.0                          23.0  
1889785         923.0                         305.0  

[1889786 rows x 7 columns]

In [209]:
small_data = aggregatedData[aggregatedData['county_code'] == 66010]

In [210]:
small_data

date_range_start  state state_code cnamelong  county_code  \
3141          2019-01-01   66.0         GU      Guam      66010.0   
6367          2019-01-02   66.0         GU      Guam      66010.0   
9593          2019-01-03   66.0         GU      Guam      66010.0   
12819         2019-01-04   66.0         GU      Guam      66010.0   
16045         2019-01-05   66.0         GU      Guam      66010.0   
...                  ...    ...        ...       ...          ...   
1876798       2020-08-04   66.0         GU      Guam      66010.0   
1880023       2020-08-05   66.0         GU      Guam      66010.0   
1883249       2020-08-06   66.0         GU      Guam      66010.0   
1886475       2020-08-07   66.0         GU      Guam      66010.0   
1889701       2020-08-08   66.0         GU      Guam      66010.0   

         device_count  completely_home_device_count  
3141           3006.0                        1560.0  
6367           2904.0                        1399.0  
9593           2925.0                        1416.0  
12819          3364.0                        1468.0  
16045          3383.0                        1650.0  
...               ...                           ...  
1876798        4712.0                        1256.0  
1880023        4750.0                        1413.0  
1883249        4882.0                        1527.0  
1886475        4703.0                        1202.0  
1889701        4757.0                        1454.0  

[586 rows x 7 columns]

So in our last dataset at this point we will get multiple rows with the same date in the same county multiple times, but now we only get 1 row for 1 county on 1 date. That is why we have so much less rows now

Now let's check if the aggregation is correct on Guam county. Let's load the data from reduced data set without going through the chunked processing. and do a groupby

In [146]:
reducedData = pd.read_csv("social_dist_all_reduced.csv", nrows = 1000000)

C:\Users\wenju\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (1,4,5,6) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [147]:
reducedData = reducedData[reducedData['county_code'] == 66010]

In [148]:
reducedData = reducedData.groupby(['date_range_start', 'state', 'state_code',\
                                             'cnamelong', 'county_code']).agg(device_count = ('device_count', 'sum'),\
                                                                              completely_home_device_count = ('completely_home_device_count', 'sum')).reset_index()

In [149]:
reducedData

date_range_start  state state_code cnamelong  county_code  \
0  2019-01-01T00:00:00+10:00   66.0         GU      Guam        66010   
1  2019-01-02T00:00:00+10:00   66.0         GU      Guam        66010   
2  2019-01-03T00:00:00+10:00   66.0         GU      Guam        66010   
3  2019-01-05T00:00:00+10:00   66.0         GU      Guam        66010   

   device_count  completely_home_device_count  
0          1842                           969  
1          1722                           810  
2          1741                           828  
3          1311                           608

Now this is super weird, this groupby without chunked processes give us device count of 1842, and the chunked process gives us 3006. Does this mean the data related to Guam county 2019-01-01 can appear later in the data? But in sequential processing shouldn't all the data related to the 2019-01-01 be in the beginning of the dataset? 

let's do an experiment, in a chunked/streaming processing fashion, get all the rows about Guam

In [156]:
GuamData = pd.read_csv("social_dist_all_reduced.csv", nrows = 0, \
                             dtype = {'date_range_start': 'datetime64[ns]',\
                                        'state': 'float64', \
                                        'state_code': 'str', \
                                        'cnamelong': 'str', \
                                        'county_code': 'float64', \
                                        'device_count': 'float64', \
                                        'completely_home_device_count': 'float64'})

In [157]:
counter = 0
for chunk in pd.read_csv("social_dist_all_reduced.csv", chunksize = 1000000):
    # get only the Guam data.
    chunk = chunk[chunk['county_code'] == 66010]
    
    # convert chunk's columns to the right datatype.
    chunk['device_count'] = pd.to_numeric(chunk['device_count'], errors='coerce')
    chunk['completely_home_device_count'] = pd.to_numeric(chunk['completely_home_device_count'], errors='coerce')
    chunk['county_code'] = pd.to_numeric(chunk['county_code'], errors='coerce')
    chunk['state'] = pd.to_numeric(chunk['state'], errors='coerce')
    # drop the null rows (get rid of the row that is header)
    # also gets rid of 2 counties whose FIPS code is null.
    chunk.dropna(inplace = True)
    # convert string column to datetime object
    chunk['date_range_start'] = chunk['date_range_start'].apply(lambda x: datetime.strptime(x[0:10], '%Y-%m-%d'))
    
    # append all the data related to Guam to GuamData
    GuamData = GuamData.append(chunk)
    
    print(counter) # track progress
    print(GuamData.shape[0]) # how large is our data right now
    counter += 1

C:\Users\wenju\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (1,4,5,6) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


0
356
1
659
2
1014
3
1304
4
1664
5
1970
6
2321
7
2580
8
2925
9
3172
10
3443
11
3780
12
4050
13
4384
14
4649
15
4961
16
5190
17
5489
18
5693
19
5927
20
6209
21
6447
22
6738
23
6959
24
7229
25
7414
26
7623
27
7743
28
7889
29
8070
30
8233
31
8416
32
8552
33
8727
34
8872
35
9066
36
9201
37
9337
38
9518
39
9649
40
9801
41
9939
42
10111
43
10237
44
10381
45
10553
46
10699
47
10889
48
11046
49
11238
50
11405
51
11591
52
11731
53
11887
54
12080
55
12246
56
12449
57
12605
58
12757
59
12871
60
13003
61
13052
62
13102
63
13159
64
13214
65
13278
66
13331
67
13400
68
13445
69
13496
70
13564
71
13602
72
13640
73
13713
74
13800
75
13861
76
14012
77
14172
78
14494
79
14883
80
15122
81
15500
82
15781
83
16169
84
16455
85
16777
86
17165
87
17425
88
17524
89
17593
90
17676
91
17738
92
17886
93
18279
94
18604
95
18977
96
19304
97
19673
98
19914
99
19992
100
20065
101
20139
102
20222
103
20293
104
20366
105
20420
106
20691
107
20956
108
21324
109
21601
110
21906
111
22292
112
22595
113
22877
114
23199
115


In [170]:
GuamData[GuamData['date_range_start'] == datetime(2019,1,1)]

date_range_start  state state_code cnamelong  county_code  \
501          2019-01-01   66.0         GU      Guam      66010.0   
1967         2019-01-01   66.0         GU      Guam      66010.0   
3576         2019-01-01   66.0         GU      Guam      66010.0   
4747         2019-01-01   66.0         GU      Guam      66010.0   
6407         2019-01-01   66.0         GU      Guam      66010.0   
...                 ...    ...        ...       ...          ...   
123530       2019-01-01   66.0         GU      Guam      66010.0   
123731       2019-01-01   66.0         GU      Guam      66010.0   
123944       2019-01-01   66.0         GU      Guam      66010.0   
125644       2019-01-01   66.0         GU      Guam      66010.0   
129566       2019-01-01   66.0         GU      Guam      66010.0   

        device_count  completely_home_device_count  
501              8.0                           5.0  
1967            11.0                           5.0  
3576            27.0                           9.0  
4747            20.0                          11.0  
6407            17.0                          12.0  
...              ...                           ...  
123530          13.0                           4.0  
123731          22.0                           9.0  
123944          20.0                           7.0  
125644           9.0                           3.0  
129566          14.0                          11.0  

[98 rows x 7 columns]

In [171]:
GuamData[GuamData['date_range_start'] == datetime(2019,1,2)]

date_range_start  state state_code cnamelong  county_code  \
264376       2019-01-02   66.0         GU      Guam      66010.0   
267084       2019-01-02   66.0         GU      Guam      66010.0   
267522       2019-01-02   66.0         GU      Guam      66010.0   
270209       2019-01-02   66.0         GU      Guam      66010.0   
272633       2019-01-02   66.0         GU      Guam      66010.0   
...                 ...    ...        ...       ...          ...   
389035       2019-01-02   66.0         GU      Guam      66010.0   
390907       2019-01-02   66.0         GU      Guam      66010.0   
391946       2019-01-02   66.0         GU      Guam      66010.0   
392155       2019-01-02   66.0         GU      Guam      66010.0   
392156       2019-01-02   66.0         GU      Guam      66010.0   

        device_count  completely_home_device_count  
264376           7.0                           5.0  
267084          36.0                          22.0  
267522          33.0                          20.0  
270209          35.0                          15.0  
272633          16.0                          11.0  
...              ...                           ...  
389035          10.0                           7.0  
390907          20.0                           7.0  
391946          12.0                           7.0  
392155          17.0                           9.0  
392156          11.0                           7.0  

[96 rows x 7 columns]

In [159]:
GuamData.groupby(['date_range_start', 'state', 'state_code',\
                  'cnamelong', 'county_code']).agg(device_count = ('device_count', 'sum'),\
                                                   completely_home_device_count = ('completely_home_device_count', 'sum')).reset_index()

date_range_start  state state_code cnamelong  county_code  device_count  \
0         2019-01-01   66.0         GU      Guam      66010.0        1842.0   
1         2019-01-02   66.0         GU      Guam      66010.0        1722.0   
2         2019-01-03   66.0         GU      Guam      66010.0        1741.0   
3         2019-01-05   66.0         GU      Guam      66010.0        1311.0   
4         2019-01-06   66.0         GU      Guam      66010.0        2086.0   
..               ...    ...        ...       ...          ...           ...   
431       2020-08-03   66.0         GU      Guam      66010.0        2994.0   
432       2020-08-05   66.0         GU      Guam      66010.0        2715.0   
433       2020-08-06   66.0         GU      Guam      66010.0        2103.0   
434       2020-08-07   66.0         GU      Guam      66010.0        2786.0   
435       2020-08-08   66.0         GU      Guam      66010.0        2704.0   

     completely_home_device_count  
0                           969.0  
1                           810.0  
2                           828.0  
3                           608.0  
4                          1029.0  
..                            ...  
431                         663.0  
432                         811.0  
433                         696.0  
434                         757.0  
435                         830.0  

[436 rows x 7 columns]

Now we have an issue.... This process of only extracting Guam data and do the groupby and aggregate gives us different device count than the one we have before.. (1842 now vs 3006 before)... Did I do something wrong? 


Let's try just doing the aggregation on the first 1 million data point

In [194]:
firstMillionData = pd.read_csv('social_dist_all_reduced.csv', nrows = 1000000)

C:\Users\wenju\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (1,4,5,6) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [198]:
firstMillionData[firstMillionData['county_code'] == '66010']

date_range_start state state_code cnamelong county_code  \
132081  2019-01-01T00:00:00+10:00  66.0         GU      Guam       66010   
135041  2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
135258  2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
137393  2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
139098  2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
...                           ...   ...        ...       ...         ...   
913885  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   
914567  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   
914807  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   
915493  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   
916398  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   

       device_count completely_home_device_count  
132081           21                           14  
135041           11                            3  
135258           21                           10  
137393           14                            6  
139098           19                            8  
...             ...                          ...  
913885           12                            3  
914567           19                            4  
914807           24                           13  
915493           31                           13  
916398           27                           20  

[401 rows x 7 columns]

In [196]:
firstMillionData[firstMillionData['county_code'] == 66010]

date_range_start state state_code cnamelong county_code  \
501     2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
1967    2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
3576    2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
4747    2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
6407    2019-01-01T00:00:00+10:00    66         GU      Guam       66010   
...                           ...   ...        ...       ...         ...   
993832  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   
994036  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   
994476  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   
994477  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   
999418  2019-01-05T00:00:00+10:00    66         GU      Guam       66010   

       device_count completely_home_device_count  
501               8                            5  
1967             11                            5  
3576             27                            9  
4747             20                           11  
6407             17                           12  
...             ...                          ...  
993832           13                            6  
994036           18                           11  
994476           25                           12  
994477           22                            7  
999418           21                            6  

[356 rows x 7 columns]

OMG now I know why our aggregation seems wrong... Because some county_code is string, and some are numbers.. This is really confusing. Let's see if merging these 2 sets together produces the right answer.

In [199]:
stringVersionData = firstMillionData[firstMillionData['county_code'] == '66010']
intVersionData = firstMillionData[firstMillionData['county_code'] == 66010]

In [200]:
mergedData = stringVersionData.append(intVersionData)

In [202]:
mergedData['device_count'] = pd.to_numeric(mergedData['device_count'], errors='coerce')
mergedData['completely_home_device_count'] = pd.to_numeric(mergedData['completely_home_device_count'], errors='coerce')
mergedData['county_code'] = pd.to_numeric(mergedData['county_code'], errors='coerce')
mergedData['state'] = pd.to_numeric(mergedData['state'], errors='coerce')

In [203]:
mergedData.groupby(['date_range_start', 'state', 'state_code',\
                    'cnamelong', 'county_code']).agg(device_count = ('device_count', 'sum'),\
                                                     completely_home_device_count = ('completely_home_device_count', 'sum')).reset_index()

date_range_start  state state_code cnamelong  county_code  \
0  2019-01-01T00:00:00+10:00   66.0         GU      Guam        66010   
1  2019-01-02T00:00:00+10:00   66.0         GU      Guam        66010   
2  2019-01-03T00:00:00+10:00   66.0         GU      Guam        66010   
3  2019-01-04T00:00:00+10:00   66.0         GU      Guam        66010   
4  2019-01-05T00:00:00+10:00   66.0         GU      Guam        66010   

   device_count  completely_home_device_count  
0          3006                          1560  
1          2904                          1399  
2          2925                          1416  
3          3364                          1468  
4          1940                           933

Okay, so now we get the device_count aggreing our aggregated data. Now we can export the aggregatedData in confidence that it is correct. **This inconsistent data schema is killing me!!!!! - some rows have string of county_code and some have int..!**

In [211]:
aggregatedData.to_csv("social_dist_aggregated_on_county.csv", index = False)